Add android support

zarik5 opened the pull request:
This is inspired from @blaind's work: https://github.com/blaind/bevy_openxr/blob/main/crates/bevy_openxr/src/platform/oculus_android/mod.rs

While @blaind work is known to be working, I didn't provide any testing for this code. If you want, I'm ok for waiting to have a working example before merging.

Ralith commented:
Thanks! An initial concern: as illustrated in the linked reference, don't we also need an `xrInitializeLoaderKHR` call?

As I'm not qualified to review JNI esoterica myself, I'd like success reports from manual testing on android at minimum.

zarik5 commented:
Yes, in the work I'm doing on bevy I'm also sending the same Java pointers to the loader using the extension. It is easy enough to do with the function pointers and it doesn't require extra modifications to openxrs. I could also contribute it directly to openxrs (inside the entry creation call), but I should definitely have something running first (which I'm deferring because I currently don't have the software + hardware ready).

zarik5 commented:
I added `Entry::initialize_android_loader()` and updated the vulkan example for android support.

This PR is now ready for review, with some caveats:

* I tested this on the Oculus Quest 1 and Quest 2 but I'm still waiting reports about the Oculus Go
* It requires cargo-apk from [this PR](https://github.com/rust-windowing/android-ndk-rs/pull/138)
* Due to the lack `khr_vulkan_enable`support, the vulkan example now has much more boilerplate.

Ralith commented on openxr/src/entry.rs:
```diff
@@ -108,7 +108,31 @@ impl Entry {
-    /// Create an OpenXR instance with certain extensions enabled
+    /// Initialize Android loader. This must be called before `create_instance()`.
+    #[cfg(target_os = "android")]
+    pub fn initialize_android_loader(&self) -> Result<()> {
```
Could we do this automatically at the start of `create_instance`?

zarik5 replied:
It would make more sense to insert this call inside `Entry::load()`, the problem is that they return different error types. While not totally correct, it is possible to insert this inside `create_instance`, and I also verified that calling this twice does not have any adverse impact.

Ralith replied:
Hmm, I'd rather not rely on coincidence there; perhaps keeping it separate is the most practical option.

Ralith commented on openxr/examples/vulkan.rs:
```diff
@@ -37,21 +40,32 @@ fn main() {
     // Initialize OpenXR with the extensions we've found!
     let mut enabled_extensions = xr::ExtensionSet::default();
-    enabled_extensions.khr_vulkan_enable2 = true;
+    enabled_extensions.khr_vulkan_enable = available_extensions.khr_vulkan_enable;
```
We should not enable version 1 if version 2 is available. Additionally, a comment should be added documenting that the fallback path is a temporary compatibility measure for Oculus runtimes.

Ralith commented on openxr/examples/vulkan.rs:
```diff
@@ -37,21 +40,32 @@ fn main() {
     let available_extensions = entry.enumerate_extensions().unwrap();
 
+    dbg!(&available_extensions);
+
```
Stray debug print?

zarik5 replied:
i'll remove it

Ralith commented on openxr/examples/vulkan.rs:
```diff
@@ -108,26 +143,85 @@ fn main() {
+                panic!("{}", res);
+            }
+            let vk_instance_exts =
+                String::from_utf8_unchecked(buffer[0..data_len as usize - 1].to_vec())
```
This should use `str::from_utf8`.

zarik5 replied:
I'll remove this code when I add proper support for `khr_vulkan_enable`.

Ralith commented on openxr/examples/vulkan.rs:
```diff
@@ -108,26 +143,85 @@ fn main() {
+                buffer.len() as _,
+                &mut data_len as _,
+                &mut buffer as *mut _ as _,
+            );
```
This (and the later call for device extensions) should use the proper two-call pattern and a `Vec` buffer to ensure sufficient space is allocated. It's fine to copy the internal helpers to accomplish this (though we might want to consider bundling them into the `sys` crate).

zarik5 replied:
I'll remove this code when I add proper support for `khr_vulkan_enable`.

Ralith reviewed (changes requested):
Thanks, left some comments. Please split the changes to the example out into a separate PR (or, if you're comfortable with history rewrites, maintain them in a separate commit) so that they can be easily reverted once Oculus updates their runtimes.

zarik5 commented on openxr/Cargo.toml:
```diff
@@ -37,3 +37,38 @@ ndk-glue = "0.3"
+    "com.oculus.intent.category.VR",
+    "android.intent.category.LAUNCHER",
+    "android.intent.category.INFO",
+]
```
This metadata applies to all `cdylib` crates and examples. Since `openxr` and `openxr-sys`are `lib`crates, they are unaffected.

Ralith replied:
Makes sense. Please add a comment to the Cargo.toml explaining this. Might also be helpful to reference it from examples/README.md as a model for how to set up a Quest app.

zarik5 commented:
I'm going to add here commits related to the implementation of the `khr_vulkan_enable` wrapper, for the sake of testing. I'll rebase later.

zarik5 commented:
The Oculus Go is probably compatible with OpenXR and this crate (I'm waiting for some last reports). The problem is that it supports only Vulkan 1.0 so the example doesn't run without heavy modifications, so I'm not listing it as compatible (but it will be compatible on bevy).

Ralith commented:
It might not be too bad to refactor the example to use the multiview extension explicitly assuming the Go implements it, but I'm just as happy to keep things more concise/modern.

zarik5 commented:
@Ralith The Oculus Go supports the extension `VK_KHR_multiview`, but adding code to support it will bloat the example even more, where it is more important showing the openxr crate usage.

zarik5 commented:
I'm ok with keeping the example on Vulkan 1.1 (without Oculus Go support). Another argument is that the cargo-apk metadata was a mashup of old and new entries for the sake of the Oculus Go support, when the user would probably decide to maintain separate Cargo.toml files.

Ralith reviewed (approved):
Thanks!

Ralith commented on openxr/examples/vulkan.rs:
```diff
@@ -47,11 +52,19 @@ fn main() {
+    } else {
+        enabled_extensions.khr_vulkan_enable = true;
+    }
+    #[cfg(target_os = "android")]
+    {
+        enabled_extensions.khr_android_create_instance = true;
+    }
```
Style nit, since more work is needed anyway:
```suggestion
    enabled_extensions.khr_android_create_instance = cfg!(target_os = "android");
```

zarik5 replied:
This doesn't work because `khr_android_create_instance` field itself is cfg-gated to Android only.

Ralith replied:
Oh, right.

Ralith commented on openxr/Cargo.toml:
```diff
@@ -32,5 +32,40 @@ ctrlc = "3.1.5"
+[[example]]
+name = "vulkan-android"
+path = "examples/vulkan.rs"
+crate-type = ["cdylib"]
```
Looks like this is causing lint issues on non-android. Can we guard the example to android-only somehow? Failing that, we should move most of the code into a common module, and add suitable conditional lint `allow`s in the vulkan-android-specific source file, perhaps combined with a `panic` with a suitable message if ran on non-android.

zarik5 replied:
I'll look into it.

zarik5 replied:
The solution was to add also the "bin" crate-type. I've seen doing this elsewhere but I wasn't aware it fixes the lints.

Ralith replied:
Huh. Does this result a useless binary being produced as well? Is the android build process unaffected?

zarik5 replied:
I just tried, the binary gets created in the target folder, but it doesn't get used. it doesn't seem to affect compilation times.

zarik5 replied:
I added a commit to use `#[allow(unused)]` instead. This is something that can be integrated in ndk_glue.

zarik5 replied:
I found a more elegant solution, make the main `pub`.

zarik5 commented:
I fixed the examples README with the crates.io link for cargo-apk. I think now it's all ready for merging.
